lib/rules/no-constant-binary-operand.js

```diff
@@ -0,0 +1,434 @@
+        case "UpdateExpression": // Numbers are not booleans
+            return true;
+        case "BinaryExpression":
+            return NUMERIC_OR_STRING_BINARY_OPERATORS.has(node.operator);
+        case "UnaryExpression": {
+            if (node.operator === "delete") {
+                return false;
+            }
+            if (node.operator === "!") {
+                return hasConstantTruthiness(node.argument);
+            }
+
+            /*
+             * The remaining operators return either strings or numbers, neither
+             * of which are boolean.
+             */
+            return true;
+        }
+        case "SequenceExpression": {
+            const last = node.expressions[node.expressions.length - 1];
+
+            return hasConstantStrictBooleanComparison(last);
+        }
+        case "Identifier":
+            return node.name === "undefined";
+        case "AssignmentExpression":
+            if (node.operator === "=") {
+                return hasConstantStrictBooleanComparison(node.right);
+            }
+
+            /*
+             * Handling short-circuting assignment operators would require
+             * walking the scope. We won't attempt that (for now...)
+             */
+            if (node.operator === "&&=" || node.operator === "||=" || node.operator === "??=") {
+                return false;
+            }
+
+            /*
+             * The remaining assignment expressions all result in either a number
+             * or a string, neither of which can ever be boolean.
+             */
+            return true;
+        case "JSXElement": // ESLint has a policy of not assuming any specific JSX behavior.
+        case "JSXFragment":
+            return false;
+        default:
+            return false;
+    }
+}
+
+
+/**
+ * Checks if one operand will cause the result to be constant.
+ * @param {ASTNode} a One side of the expression
+ * @param {ASTNode} b The other side of the expression
+ * @param {string} operator The binary expression operator
+ * @returns {ASTNode | null} The node which will cause the expression to have a constant result.
+ */
+function findBinaryExpressionConstantOperand(a, b, operator) {
+    if (operator === "==" || operator === "!=") {
+        if (
+            (isNullOrUndefined(a) && hasConstantNullishness(b)) ||
+            (isStaticBoolean(a) && hasConstantLooseBooleanComparison(b))
+        ) {
+            return b;
+        }
+    } else if (operator === "===" || operator === "!==") {
+        if (
+            (isNullOrUndefined(a) && hasConstantNullishness(b)) ||
+            (isStaticBoolean(a) && hasConstantStrictBooleanComparison(b))
+        ) {
+            return b;
+        }
+    }
+    return null;
+}
+
+
+//------------------------------------------------------------------------------
+// Rule Definition
+//------------------------------------------------------------------------------
+
+module.exports = {
```

Comment:
There's a new comment that should be added here per https://github.com/eslint/eslint/pull/15291:

```suggestion
/** @type {import('eslint').Rule.RuleModule} */
module.exports = {
```

Reply:
Looks like it ended up needing to be `/** @type {import('../shared/types').Rule} */`.